packages/block-editor/src/components/link-control/index.js

```diff
@@ -0,0 +1,236 @@
+/**
+ * External dependencies
+ */
+import classnames from 'classnames';
+import { isFunction, noop, startsWith } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import {
+	Button,
+	ExternalLink,
+	Popover,
+} from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
+
+import {
+	useCallback,
+	useState,
+	useEffect,
+	Fragment,
+} from '@wordpress/element';
+
+import {
+	safeDecodeURI,
+	filterURLForDisplay,
+	isURL,
+	prependHTTP,
+	getProtocol,
+} from '@wordpress/url';
+
+import { withInstanceId } from '@wordpress/compose';
+
+/**
+ * Internal dependencies
+ */
+import LinkControlSettingsDrawer from './settings-drawer';
+import LinkControlSearchItem from './search-item';
+import LinkControlInputSearch from './input-search';
+
+function LinkControl( {
+	instanceId,
+	currentLink,
+	className,
+	fetchSearchSuggestions,
+	onLinkChange,
+	currentSettings,
+	onSettingsChange = { noop },
+	onKeyDown = noop,
+	onKeyPress = noop,
+	onClose = noop,
+} ) {
+	// State
+	const [ inputValue, setInputValue ] = useState( '' );
+	const [ isEditingLink, setIsEditingLink ] = useState( false );
+
+	// Effects
+	useEffect( () => {
+		// If we have a link then stop editing mode
+		if ( currentLink ) {
+			setIsEditingLink( false );
+		} else {
+			setIsEditingLink( true );
+		}
+	}, [ currentLink ] );
+
+	// Handlers
+	const onInputChange = ( value = '' ) => {
+		setInputValue( value );
+	};
+
+	const onLinkSelect = ( suggestion ) => ( event ) => {
+		event.preventDefault();
+		event.stopPropagation();
+
+		if ( isFunction( onLinkChange ) ) {
+			onLinkChange( suggestion );
+		}
+	};
+
+	// Utils
+
+	const startEditMode = () => {
+		setIsEditingLink( true );
+	};
+
+	const closeLinkUI = () => {
+		resetInput();
+		onClose();
+	};
+
+	const resetInput = () => {
+		setInputValue( '' );
+	};
+
+	const handleDirectEntry = async ( value ) => {
```

Comment:
This is still marked as async, yet isn't an async function.

Reply:
Ah yes. I've now utilised `Promise.resolve` to ensure we return a `Promise` from this function. We need this because the handler inside `URLInput` expects a `Promise` and I don't want to mess with any more internals of that component.